field/internal/templates/element/base.go

```diff
@@ -433,23 +463,43 @@ func _addGeneric(z,  x, y *{{.ElementName}}) {
 		{{- $hasCarry := or (not $.NoCarry) (lt $i $.NbWordsLastIndex)}}
 		z[{{$i}}], {{- if $hasCarry}}carry{{- else}}_{{- end}} = bits.Add64(x[{{$i}}], y[{{$i}}], {{- if eq $i 0}}0{{- else}}carry{{- end}})
 	{{- end}}
-	{{- if not .NoCarry}}
-		// if we overflowed the last addition, z >= q
-		// if z >= q, z = z - q
-		if carry != 0 {
-			// we overflowed, so z >= q
-			z[0], {{- if gt $.NbWords 1}}carry{{- else}}_{{- end}} = bits.Sub64(z[0], {{index $.Q 0}}, 0)
-			{{- range $i := .NbWordsIndexesNoZero}}
-				z[{{$i}}], carry = bits.Sub64(z[{{$i}}], {{index $.Q $i}}, carry)
-			{{- end}}
-			return
+
+	{{- if eq $.NbWords 1}}
+		if {{- if not .NoCarry}} carry != 0 ||{{- end }} z[0] >= q {
+			z[0] -= q
 		}
-	{{- end}}
+	{{- else}}
+		{{- if not .NoCarry}}
+			// if we overflowed the last addition, z >= q
+			// if z >= q, z = z - q
+			if carry != 0 {
+				var b uint64
+				// we overflowed, so z >= q
+				{{- range $i := iterate 0 $.NbWords}}
+					{{- $hasBorrow := lt $i $.NbWordsLastIndex}}
+					z[{{$i}}], {{- if $hasBorrow}}b{{- else}}_{{- end}} = bits.Sub64(z[{{$i}}], {{index $.Q $i}}, {{- if eq $i 0}}0{{- else}}b{{- end}})
+				{{- end}}
+				return
+			}
+		{{- end}}
 
-	{{ template "reduce" .}}
+		{{ template "reduce" .}}
```

Comment:
Should `reduce` be in an else clause? There's no situation where we would need two reductions. Does it matter?

Reply:
I don't understand the two reductions part / does it matter --> both code path must exist at the template level, when all bits are set and you have a carry to the last addition, then you need to reduce, but if not, you still need to check that `z < q`.